### PR TITLE
Users/shreyas r msft/mocha test result parser

### PR DESCRIPTION
**Pull request by ShreyasRmsft**

### CODE REVIEW

**Comment by singhsarab** [comment]
> namespace Agent.Plugins.TestResultParser.Loggers.Interfaces

Copyright

---
Refers to: src/Agent.Plugins/Loggers/Interfaces/IDiagnosticDebugLogger.cs:1 in bfa64ed. [](commit_id = bfa64ed8c940e44a2d6b1a5b5116c5360d04c5c7, deletion_comment = False)

File: src/Agent.Plugins/TestResult/Models/TestRun.cs [hunk 1]
```diff
@@ -10,6 +10,16 @@ namespace Agent.Plugins.TestResultParser.TestResult.Models
     /// </summary>
     public class TestRun
     {
+        /// <summary>
+        /// Name of the parser class that parsed the test run
+        /// </summary>
+        public string Parser { get; set; }
```
Comment (singhsarab) [inline src/Agent.Plugins/TestResult/Models/TestRun.cs]: Consider URI

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs [hunk 2]
```diff
@@ -1,19 +1,597 @@
-﻿namespace Agent.Plugins.TestResultParser.Parser.Node.Mocha
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+// TODO: Add heavy verbose logging 
+
+namespace Agent.Plugins.TestResultParser.Parser.Node.Mocha
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: TODO: TestRun Id, Parser URI

File: src/Agent.Plugins/TestResult/Models/TestRunSummary.cs [hunk 3]
```diff
@@ -21,7 +21,12 @@ public class TestRunSummary
         /// <summary>
         /// Total number of tests that were passed
         /// </summary>
-        public int TotalPassed { get; set; }
+        public int TotalPassed { get; set; }
+
+        /// <summary>
+        /// Total number of tests that were skipped
+        /// </summary>
+        public int TotalSkipped { get; set; }
```
Comment (singhsarab) [inline src/Agent.Plugins/TestResult/Models/TestRunSummary.cs]: >TotalSkipped [](start = 19, length = 12)

Add validation part in the TestRunManager

**Comment by singhsarab** [comment]
> <Project Sdk="Microsoft.NET.Sdk">

Rename this csproj and folder, and fix the namespaces

---
Refers to: tests/Agent.Plugins.UnitTests/Agent.Plugins.UnitTests/Agent.Plugins.UnitTests.csproj:1 in bfa64ed. [](commit_id = bfa64ed8c940e44a2d6b1a5b5116c5360d04c5c7, deletion_comment = False)

File: src/Agent.Plugins/Gateway/DataStreamGateway.cs [hunk 4]
```diff
@@ -10,15 +10,17 @@
...
 
 namespace Agent.Plugins.TestResultParser.Gateway
 {
     public class DataStreamGateway : IDataStreamGateway, IBus<LogLineData>
     {
         public void Initialize()
         {
-            var t = new GenericTestResultParser();
-            Subscribe(t.Parse);
+            //var t = new MochaTestResultParser(new TestRunManager());
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Gateway/DataStreamGateway.cs]: will do.

File: src/Agent.Plugins/Loggers/DiagnosticDataCollector.cs [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
+        private static ITraceLogger instance;
 
         /// <summary>
         /// Gets the singleton instance of diagnostics data collector.
         /// </summary>
-        public static IDiagnosticDataCollector Instance
+        public static ITraceLogger Instance
         {
-            get => instance ?? (instance = new DiagnosticDataCollector());
+            get => instance ?? (instance = new TraceLogger());
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Loggers/DiagnosticDataCollector.cs]: was already written code, just renamed it, Will refactor it though

File: src/Agent.Plugins/Loggers/DiagnosticDataCollector.cs [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
             internal set => instance = value;
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Loggers/DiagnosticDataCollector.cs]: was already written code, just renamed it, Will refactor it though

File: src/Agent.Plugins/Parser/Interfaces/ITestResultParser.cs [hunk 7]
```diff
@@ -1,13 +1,23 @@
...
+        /// <summary>
+        /// Parse task output line by line to detect the test result
+        /// </summary>
+        /// <param name="line">Data to be parsed.</param>
+        void Parse(LogLineData line);
+
+        /// <summary>
+        /// Name of the parser
+        /// </summary>
+        string ParserName { get; }
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/Interfaces/ITestResultParser.cs]: structure in the interface? not sure what that means can you please elaborate?

File: src/Agent.Plugins/Parser/RegexCaptureGroups.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Agent.Plugins.TestResultParser.Parser
+{
+    public class RegexCaptureGroups
+    {
+        public static string TestCaseName { get; } = "TestCaseName";
```
[the remark this answers is not in the document]
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/RegexCaptureGroups.cs]: this was to allow overriding from the server config, i would only need to add a setter then.

But if we decide we are not doing that then I can change to const

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs [hunk 9]
```diff
@@ -1,19 +1,507 @@
...
+
+                // If no failed tests found then skip the stack traces parsing state
+                if (testRun.TestRunSummary.TotalFailed == 0)
+                {
+                    state = MochaTestResultParserState.ExpectingTestResults;
+                    PublishRunAndResetParser();
+                }
+                else
+                {
+                    state = MochaTestResultParserState.ExpectingStackTraces;
```
Comment (cltshivash) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: >state = MochaTestResultParserState.ExpectingStackTraces; [](start = 20, length = 56)

transition logs are important..

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs (hunk 2 above)
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: done

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs (hunk 9 above)
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: done

File: src/Agent.Plugins/TestResult/Models/TestRunSummary.cs (hunk 3 above)
Comment (ShreyasRmsft) [reply src/Agent.Plugins/TestResult/Models/TestRunSummary.cs]: next pr, dont want to bloat this one

File: src/Agent.Plugins/Loggers/TraceLogger.cs [hunk 10]
```diff
@@ -30,9 +30,9 @@ public void Info(string text)
         }
 
         /// <inheritdoc />
-        public Task PublishDiagnosticDataAsync()
-        {
-            throw new NotImplementedException();
+        public Task PublishDiagnosticDataAsync()
```
Comment (nigurr) [inline src/Agent.Plugins/Loggers/TraceLogger.cs]: fix

File: src/Agent.Plugins/Parser/Interfaces/ITestResultParser.cs [hunk 11]
```diff
@@ -1,13 +1,23 @@
...
+
+        /// <summary>
+        /// Version number of the parser
+        /// </summary>
+        string ParserVersion { get; }
```
Comment (nigurr) [inline src/Agent.Plugins/Parser/Interfaces/ITestResultParser.cs]: Version

File: src/Agent.Plugins/Parser/Models/LogLineData.cs [hunk 12]
```diff
@@ -3,5 +3,7 @@
     public class LogLineData
```
Comment (nigurr) [inline src/Agent.Plugins/Parser/Models/LogLineData.cs]: LineData

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserState.cs [hunk 13]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Agent.Plugins.TestResultParser.Parser.Node.Mocha
```
Comment (nigurr) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserState.cs]: fixing ns

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs [hunk 14]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Agent.Plugins.TestResultParser.Parser.Node.Mocha
+{
+    public class MochaTestResultParserStateContext : ParserStateContextBase
+    {
+        public int StackTracesToSkipParsingPostSummary { get; set; } = 0;
```
Comment (nigurr) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs]: write summary

File: src/Agent.Plugins/Parser/ParserStateContextBase.cs [hunk 15]
```diff
@@ -0,0 +1,10 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Agent.Plugins.TestResultParser.Parser
+{
+    public abstract class ParserStateContextBase
```
Comment (nigurr) [inline src/Agent.Plugins/Parser/ParserStateContextBase.cs]: delete it

File: src/Agent.Plugins/Parser/RegexCaptureGroups.cs [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public class RegexCaptureGroups
+    {
+        public static string TestCaseName { get; } = "TestCaseName";
+
+        public static string FailedTestCaseNumber { get; } = "FailedTestCaseNumber";
+
+        public static string PassedTests { get; } = "PassedTests";
+
+        public static string FailedTests { get; } = "FailedTests";
```
Comment (nigurr) [inline src/Agent.Plugins/Parser/RegexCaptureGroups.cs]: const

File: src/Agent.Plugins/Telemetry/TelemetryDataCollector.cs [hunk 17]
```diff
@@ -13,12 +13,12 @@ public class TelemetryDataCollector : ITelemetryDataCollector
...
-
-        /// <inheritdoc />
-        public void AddProperty(string key, object value, bool aggregate = false)
-        {
-            throw new NotImplementedException();
-        }
+        }
+
+        /// <inheritdoc />
+        public void AddToCumulativeTelemtery(string EventArea, string EventName, object value, bool aggregate = false)
```
Comment (nigurr) [inline src/Agent.Plugins/Telemetry/TelemetryDataCollector.cs]: ??

**Pull request merged**